### PR TITLE
Proposed Feature: Add DatabaseTasks#connection_class to simplify multiple database apps

### PR DESCRIPTION
**Pull request by bf4**
### Summary

Replacing `ActiveRecord::Base` with a `connection_class` method allows easier re-use of the
DatabaseTasks object in an app that has connections to multiple databases.

### Other Information

There may be other places where we'd want to reduce the `ActiveRecord::Base` 'obsession', but I've omitted them from this PR to focus on the concept of whether this is a pattern Rails-core is interested in.  I create in in response to having to write tasks which cache the current ActiveRecord::Base connection_config, establish a connection to a different database, run the task, then restore the original connection config.

### CODE REVIEW

**Comment by rafaelfranca** [comment]
👍  for the idea. I was planing to implement it myself and we have support to multiple databases in our plans for a while now. But is not this already possible with the `database_configuration` option?

File: activerecord/lib/active_record/tasks/database_tasks.rb [hunk 1]
```diff
@@ -219,7 +219,7 @@ def purge_current(environment = env)
         each_current_configuration(environment) { |configuration|
           purge configuration
         }
-        ActiveRecord::Base.establish_connection(environment.to_sym)
+        Aconnection_class.establish_connection(environment.to_sym)
```
Comment (r-obert) [inline activerecord/lib/active_record/tasks/database_tasks.rb]: -A 😃

**Comment by bf4** [comment]
@rafaelfranca 

> Is not this already possible with the `ActiveRecord::Base.configurations` option?

It may be and I just haven't come across code that suggested this would help and so hadn't considered it. Is there a notion of a 'scoped' active configuration that allows ActiveRecord::Base.connection to be different within the task?  

<details>
  <summary>
(By 'scoped', I exclude workarounds like the below. My point is just that it is a lot of non-obvious code. Click to expand)
  </summary>

```ruby
# Taken from http://www.ostinelli.net/setting-multiple-databases-rails-definitive-guide/
# Written at Rails 4.2
namespace :postgres do
  namespace :db do |ns|
    task :drop do
      connection_config = @connection_config
      ActiveRecord::Tasks::DatabaseTasks.drop(connection_config)
    end

    task :create do
      connection_config = @connection_config
      ActiveRecord::Tasks::DatabaseTasks.create(connection_config)
    end

    task setup: ['postgres:db:create', 'postgres:db:structure:load', 'postgres:db:seed']

    task :migrate do
      Rake::Task[:environment].invoke
      Rake::Task['db:load_config'].invoke
      begin
        verbose = ENV["VERBOSE"] ? ENV["VERBOSE"] == "true" : true
        version = ENV["VERSION"] ? ENV["VERSION"].to_i : nil
        scope   = ENV['SCOPE']
        verbose_was = ActiveRecord::Migration.verbose
        ActiveRecord::Migration.verbose = verbose
        migrations_paths = Rails.application.paths['db/migrate'].to_a
        ActiveRecord::Migrator.migrate(migrations_paths, version) do |migration|
          scope.blank? || scope == migration.scope
        end
        ActiveRecord::Base.clear_cache!
      ensure
        ActiveRecord::Migration.verbose = verbose_was
      end
      Rake::Task['postgres:db:structure:dump'].invoke
    end

    namespace :migrate do
      task :down do
        Rake::Task["db:migrate:down"].invoke
      end

      task :up do
        Rake::Task["db:migrate:up"].invoke
      end
    end

    task :reset do
      Rake::Task["db:reset"].invoke
    end

    task :rollback do
      Rake::Task["db:rollback"].invoke
    end

    task :seed do
      Rails.application.config.paths["db/seeds"].each { |file| load(file) }
    end

    task :version do
      Rake::Task["db:version"].invoke
    end

    namespace :structure do
      # Structure tasks call pg_dump and psql to dump/load the structure
      # respectively. Those tasks don't know how to parse our configs, so we
      # override them here.

      task :load do
        connection_config = @connection_config
        ActiveRecord::Tasks::DatabaseTasks.structure_load(connection_config, ENV["DB_STRUCTURE"])
      end

      task dump: [:environment] do
        next if Rails.application.config.active_record.dump_schema_after_migration == false
        filename = ENV['DB_STRUCTURE'] || File.join(@db_dir, "structure.sql")

        connection_config = @connection_config
        ActiveRecord::Tasks::DatabaseTasks.structure_dump(connection_config, filename)

        if ActiveRecord::Base.connection.supports_migrations? &&
            ActiveRecord::SchemaMigration.table_exists?
          File.open(filename, "a") do |f|
            f.puts ActiveRecord::Base.connection.dump_schema_information
            f.print "\n"
          end
        end
      end
    end

    namespace :test do
      task :prepare do
        Rake::Task["db:test:prepare"].invoke
      end
    end

    # append and prepend proper tasks to all the tasks defined here above
    ns.tasks.each do |task|
      task.enhance ["postgres:set_custom_config"] do
        # If multiple tasks from this file are run as a chain (e.g.
        # postgres:db:setup), we need to ensure that each invocation actually
        # executes set_custom_config. To do this, we explicitly reenable it.
        Rake::Task["postgres:set_custom_config"].reenable
        Rake::Task["postgres:revert_to_original_config"].execute
      end
    end
  end

  task :set_custom_config do
    # save current vars
    @original_config = {
      env_db_structure: ENV["DB_STRUCTURE"],
      config: Rails.application.config.dup
    }

    # set config variables for custom database
    ENV["DB_STRUCTURE"] = "db/postgres/structure.sql"
    Rails.application.config.paths["db"] = ["db/postgres"]
    Rails.application.config.paths["db/migrate"] = ["db/postgres/migrate"]
    Rails.application.config.paths["db/seeds"] = ["db/postgres/seeds.rb"]
    Rails.application.config.paths["config/database"] = ["config/postgres.yml"]
    database_configuration = YAML.load(ERB.new(Rails.root.join('config', 'postgres.yml').read).result).with_indifferent_access # rubocop:disable Security/YAMLLoad
    @original_database_configuration = ActiveRecord::Tasks::DatabaseTasks.database_configuration || {}
    (ActiveRecord::Tasks::DatabaseTasks.database_configuration || {}).clear
    ActiveRecord::Tasks::DatabaseTasks.database_configuration = database_configuration
    @original_migrations_paths = ActiveRecord::Tasks::DatabaseTasks.migrations_paths
    ActiveRecord::Tasks::DatabaseTasks.migrations_paths.clear
    ActiveRecord::Tasks::DatabaseTasks.migrations_paths << 'db/postgres/migrate'
    @connection_config = database_configuration[Rails.env].except(:options)
    # Exclude options:
    # If we had default-transaction-read-only=true, the db setup rake task would raise
    #
    # ```
    # PG::ReadOnlySqlTransaction: ERROR:
    #  cannot execute CREATE DATABASE in a read-only transaction:
    #  CREATE DATABASE "app_test" ENCODING = 'utf8'
    # ```
    ActiveRecord::Base.establish_connection @connection_config
  end

  task :revert_to_original_config do
    # reset config variables to original values
    ENV["DB_STRUCTURE"] = @original_config[:env_db_structure]
    Rails.application.config = @original_config[:config]
    ActiveRecord::Tasks::DatabaseTasks.database_configuration.clear
    ActiveRecord::Tasks::DatabaseTasks.database_configuration.update(@original_database_configuration)
    ActiveRecord::Tasks::DatabaseTasks.migrations_paths.clear
    ActiveRecord::Tasks::DatabaseTasks.migrations_paths.concat @original_migrations_paths
    ActiveRecord::Base.connection.disconnect!
    ActiveRecord::Base.establish_connection Rails.application.config_for(:database)
  end
end
```

</details>

**Comment by eileencodes** [comment]
Hi @bf4! I'm working on some improvements to multiple databases as well. Would your problem be solved if Rails made the rake tasks for you (something I'm currently working on)? Or would you still benefit from the `connection_class`. I'm trying to better understand your use case.

**Comment by bf4** [comment]
@eileencodes (sorry for late reply; the notification got buried).  It would definitely help if Rails made the rake tasks for me, as I wouldn't have come down this road in the first place, I don't think.  I would imagine these changes to be useful to any use of the class.   In my experience, making the connection class configurable and not just `ActiveRecord::Base` avoids a lot of problems here and in other gems :)  I can't really say more without looking at they kinds of things your doing. If it's not public, yet, I'm in the basecamp.

**Comment by eileencodes** [comment]
Cool ok that makes sense to me. One more question - it wasn't clear to me where I as a user of Rails would set the `connection_class` so that the rake tasks know which connection to use. Do you have an example of how you'd use this in your app?

**Comment by bf4** [comment]
@eileencodes So, in our app, we made a `db_postgres.rake` task

our database.yml refers to our mysql db

we added a `postgres.yml`

and our rake task was

```ruby
# Taken from http://www.ostinelli.net/setting-multiple-databases-rails-definitive-guide/
task spec: ["postgres:db:test:prepare"]

namespace :postgres do
  namespace :db do |ns|

    # append and prepend proper tasks to all the tasks defined here above
    ns.tasks.each do |task|
      task.enhance ["postgres:set_custom_config"] do
        # If multiple tasks from this file are run as a chain (e.g.
        # postgres:db:setup), we need to ensure that each invocation actually
        # executes set_custom_config. To do this, we explicitly reenable it.
        Rake::Task["postgres:set_custom_config"].reenable
        Rake::Task["postgres:revert_to_original_config"].execute
      end
    end
  end
end
```

```ruby
  task :set_custom_config do
    # save current vars
    @original_config = {
      env_db_structure: ENV["DB_STRUCTURE"],
      config: Rails.application.config.dup
    }

    # set config variables for custom database
    ENV["DB_STRUCTURE"] = "db/postgres/structure.sql"
    Rails.application.config.paths["db"] = ["db/postgres"]
    Rails.application.config.paths["db/migrate"] = ["db/postgres/migrate"]
    Rails.application.config.paths["db/seeds"] = ["db/postgres/seeds.rb"]
    Rails.application.config.paths["config/database"] = ["config/postgres.yml"]
    database_configuration = YAML.load(ERB.new(Rails.root.join('config', 'postgres.yml').read).result).with_indifferent_access # rubocop:disable Security/YAMLLoad
    @original_database_configuration = ActiveRecord::Tasks::DatabaseTasks.database_configuration || {}
    (ActiveRecord::Tasks::DatabaseTasks.database_configuration || {}).clear
    ActiveRecord::Tasks::DatabaseTasks.database_configuration = database_configuration
    @original_migrations_paths = ActiveRecord::Tasks::DatabaseTasks.migrations_paths
    ActiveRecord::Tasks::DatabaseTasks.migrations_paths.clear
    ActiveRecord::Tasks::DatabaseTasks.migrations_paths << 'db/postgres/migrate'
    @connection_config = database_configuration[Rails.env].except(:options)
    # Exclude options:
    # If we had default-transaction-read-only=true, the db setup rake task would raise
    #
    # ```
    # PG::ReadOnlySqlTransaction: ERROR:
    #  cannot execute CREATE DATABASE in a read-only transaction:
    #  CREATE DATABASE "myapp_test" ENCODING = 'utf8'
    # ```
    ActiveRecord::Base.establish_connection @connection_config
  end

  task :revert_to_original_config do
    # reset config variables to original values
    ENV["DB_STRUCTURE"] = @original_config[:env_db_structure]
    Rails.application.config = @original_config[:config]
    ActiveRecord::Tasks::DatabaseTasks.database_configuration.clear
    ActiveRecord::Tasks::DatabaseTasks.database_configuration.update(@original_database_configuration)
    ActiveRecord::Tasks::DatabaseTasks.migrations_paths.clear
    ActiveRecord::Tasks::DatabaseTasks.migrations_paths.concat @original_migrations_paths
    ActiveRecord::Base.connection.disconnect!
    ActiveRecord::Base.establish_connection Rails.application.config_for(:database)
  end
```

which could have been

```ruby
PostgresDbTask = Class.new(ActiveRecord::Tasks::DatabaseTask)
```

and for completeness, here's the rest

```ruby
    task :drop do
      connection_config = @connection_config
      ActiveRecord::Tasks::DatabaseTasks.drop(connection_config)
    end

    task :create do
      connection_config = @connection_config
      ActiveRecord::Tasks::DatabaseTasks.create(connection_config)
    end

    task setup: ['postgres:db:create', 'postgres:db:structure:load', 'postgres:db:seed']

    task :migrate do
      Rake::Task[:environment].invoke
      Rake::Task['db:load_config'].invoke
      begin
        verbose = ENV["VERBOSE"] ? ENV["VERBOSE"] == "true" : true
        version = ENV["VERSION"] ? ENV["VERSION"].to_i : nil
        scope   = ENV['SCOPE']
        verbose_was = ActiveRecord::Migration.verbose
        ActiveRecord::Migration.verbose = verbose
        migrations_paths = Rails.application.paths['db/migrate'].to_a
        ActiveRecord::Migrator.migrate(migrations_paths, version) do |migration|
          scope.blank? || scope == migration.scope
        end
        ActiveRecord::Base.clear_cache!
      ensure
        ActiveRecord::Migration.verbose = verbose_was
      end
      Rake::Task['postgres:db:structure:dump'].invoke
    end

    namespace :migrate do
      task :down do
        Rake::Task["db:migrate:down"].invoke
      end

      task :up do
        Rake::Task["db:migrate:up"].invoke
      end
    end

    task :reset do
      Rake::Task["db:reset"].invoke
    end

    task :rollback do
      Rake::Task["db:rollback"].invoke
    end

    task :seed do
      Rails.application.config.paths["db/seeds"].each { |file| load(file) }
    end

    task :version do
      Rake::Task["db:version"].invoke
    end

    namespace :structure do
      # Structure tasks call pg_dump and psql to dump/load the structure
      # respectively. Those tasks don't know how to parse makara configs, so we
      # override them here.

      task :load do
        connection_config = @connection_config
        ActiveRecord::Tasks::DatabaseTasks.structure_load(connection_config, ENV["DB_STRUCTURE"])
      end

      task dump: [:environment] do
        next if Rails.application.config.active_record.dump_schema_after_migration == false
        filename = ENV['DB_STRUCTURE'] || File.join(@db_dir, "structure.sql")

        connection_config = @connection_config
        ActiveRecord::Tasks::DatabaseTasks.structure_dump(connection_config, filename)

        if ActiveRecord::Base.connection.supports_migrations? &&
            ActiveRecord::SchemaMigration.table_exists?
          File.open(filename, "a") do |f|
            f.puts ActiveRecord::Base.connection.dump_schema_information
            f.print "\n"
          end
        end
      end
    end

    namespace :test do
      task :prepare do
        Rake::Task["db:test:prepare"].invoke
      end
    end
```

**Comment by jrochkind** [comment]
If this is still relevant, does it make sense to rebase to avoid conflicts, and then maybe merge?

**Comment by bf4** [comment]
@jrochkind If there's interest I'd rebase.

**Comment by eileencodes** [comment]
I think since we're doing so much work on Rails regarding multiple databases that it will either make this unnecessary or make this PR un-rebasable. Let's see where we're at closer to 6 and whether this will still be necessary. 

Thanks for taking the time to work on this!

**Comment by bf4** [comment]
Sure. Let me know if I can help in any way.   Thanks for your work.

**Comment by bf4** [comment]
related https://github.com/rails/rails/pull/34052

**Pull request closed without merging**